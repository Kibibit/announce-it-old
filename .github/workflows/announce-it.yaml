name: Announce Release

on:
  workflow_run:
    workflows: [ "Release" ]
    branches: [ master ]
    types: 
      - completed

jobs:
  announce:
    runs-on: ubuntu-18.04
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Setup Node.js
        uses: actions/setup-node@v1
        with:
          node-version: 12
      - name: Install dependencies
        run: |
          npm ci
          npm install
          npm run build
      - name: announce-it
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
          TWITTER_ACCESS_TOKEN_KEY: ${{ secrets.TWITTER_ACCESS_TOKEN_KEY }}
          TWITTER_ACCESS_TOKEN_SECRET: ${{ secrets.TWITTER_ACCESS_TOKEN_SECRET }}
          TWITTER_CONSUMER_KEY: ${{ secrets.TWITTER_CONSUMER_KEY }}
          TWITTER_CONSUMER_SECRET: ${{ secrets.TWITTER_CONSUMER_SECRET }}
        run: npm run start:ci