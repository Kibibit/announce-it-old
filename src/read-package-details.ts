import fs from 'fs-extra';
import { get, isString } from 'lodash';
import { IBranchObject } from './announce-it';

export interface IPackageDetails {
  name: string;
  version: string;
  description: string;
  author: string;
  homepage: string;
  repository: {
    type: string;
    url: string;
  };
  announcements: {
    tweet: string;
    includeUnstable?: boolean;
  };
  release: {
    branches: Array<string | IBranchObject>;
  }
}

export function readPackageDetails(root: string): Promise<IPackageDetails> {
  return fs.readJson(`${ root }/package.json`)
    .then((packageDetails: Partial<IPackageDetails>) => {
      if (!isString(get(packageDetails, 'announcements.tweet'))) {
        throw new Error('no tweet template found. please see the readme for more details');
      }

      // here, we KNOW the package has the announcement template string
      return packageDetails as IPackageDetails;
    });
}
