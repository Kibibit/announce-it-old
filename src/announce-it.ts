import { forEach, get, isNil, isObject, template } from 'lodash';
import Twitter from 'twitter-lite';

import { IPackageDetails } from './read-package-details';

export interface IKbAnnounceItOptions {
  consumerKey: string;
  consumerSecret: string;
  accessTokenKey: string;
  accessTokenSecret: string;
}

export class KbAnnounceIt {
  private client: any;
  constructor(options: IKbAnnounceItOptions) {

    if (!get(options, 'consumerKey') ||
      !get(options, 'consumerSecret') ||
      !get(options, 'accessTokenKey') ||
      !get(options, 'accessTokenSecret')) {

      throw new Error('ERROR: missing required options');
    }

    this.client = new Twitter({
      subdomain: 'api',
      consumer_key: options.consumerKey,
      consumer_secret: options.consumerSecret,
      access_token_key: options.accessTokenKey,
      access_token_secret: options.accessTokenSecret
    });
  }

  announceRelease(packageDetails: IPackageDetails): Promise<string> {
    if (
      !packageDetails.announcements.includeUnstable &&
      !packageDetails.version.match(/^(\d+\.)+\d+$/)
    ) {
      return Promise.reject(new Error('Not a stable release'));
    }

    const tweet = this.generateTweet(packageDetails);

    return this.client
      .get('account/verify_credentials')
      .then(() => this.client.post('statuses/update', { status: tweet }))
      .then(() => tweet);
  }

  generateTweet(packageDetails: IPackageDetails): string {
    this.ensureKeyAttributes(packageDetails);

    const tweetTemplate = template(packageDetails.announcements.tweet);
    let tweet: string;
    try {
      tweet = tweetTemplate({
        package: packageDetails.name,
        version: packageDetails.version,
        description: packageDetails.description,
        author: packageDetails.author,
        homepage: packageDetails.homepage,
        npmpage: `https://www.npmjs.com/package/${ packageDetails.name }/v/${ packageDetails.version }`
      });
    } catch (e) {
      throw new Error('Using missing variables in tweet template');
    }

    return tweet;
  }

  private ensureKeyAttributes(packageDetails: Partial<IPackageDetails>): void {
    const requiredKeys = [ 'name', 'version', 'announcements.tweet' ];

    if (!isObject(packageDetails)) {
      throw new Error('Expecting Object');
    }

    forEach(requiredKeys, (key) => {
      if (isNil(get(packageDetails, key))) {
        throw new Error(`The key ${ key } is missing from given object`);
      }
    });
  }
}
