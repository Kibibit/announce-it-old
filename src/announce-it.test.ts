import { KbAnnounceIt } from './announce-it';
import { PackageDetails } from './read-package-details';

const TwitterMocks = () => {
  const defaultMock = (...args: any) => Promise.resolve();

  let getMock = defaultMock;
  let postMock = defaultMock;

  return {
    mock: {
      get: () => getMock(),
      post: () => postMock()
    },
    reset: (userDefinedMocks: any = {}) => {
      getMock = userDefinedMocks.get || getMock;
      postMock = userDefinedMocks.post || postMock;
    }
  }
};

const twitterMocks = TwitterMocks();

jest.mock('twitter-lite', () => () => twitterMocks.mock);

describe('KbAnnounceIt', () => {
  it('should throw error when missing options', () => {
    expect.assertions(5);

    // @ts-ignore
    expect(() => new KbAnnounceIt()).toThrow(Error);

    // @ts-ignore
    expect(() => new KbAnnounceIt({
      accessTokenSecret: 'TEST',
      consumerKey: 'TEST',
      consumerSecret: 'TEST'
    })).toThrow(Error);

    // @ts-ignore
    expect(() => new KbAnnounceIt({
      accessTokenKey: 'TEST',
      consumerKey: 'TEST',
      consumerSecret: 'TEST'
    })).toThrow(Error);

    // @ts-ignore
    expect(() => new KbAnnounceIt({
      accessTokenKey: 'TEST',
      accessTokenSecret: 'TEST',
      consumerSecret: 'TEST'
    })).toThrow(Error);
    // @ts-ignore
    expect(() => new KbAnnounceIt({
      accessTokenKey: 'TEST',
      accessTokenSecret: 'TEST',
      consumerKey: 'TEST'
    })).toThrow(Error);

  });

  it('should create instance when given correct parameters', () => {
    const announceIt = new KbAnnounceIt({
      accessTokenKey: 'TEST',
      accessTokenSecret: 'TEST',
      consumerKey: 'TEST',
      consumerSecret: 'TEST'
    });

    expect(announceIt).toBeDefined();
  });
});

describe('kbAnnounceIt.announceRelease', () => {
  const packageDetails: PackageDetails = {
    name: 'test-repo',
    description: 'test-description',
    author: 'test-author',
    homepage: 'test-homepage',
    repository: {
      type: 'test-repo-type',
      url: 'test-repo-url'
    },
    version: 'test-version',
    announcements: {
      tweet: 'test-template'
    }
  };

  let announceIt: KbAnnounceIt;

  beforeEach(() => {
    announceIt = new KbAnnounceIt({
      accessTokenKey: 'TEST',
      accessTokenSecret: 'TEST',
      consumerKey: 'TEST',
      consumerSecret: 'TEST'
    });
  });

  it('should throw error on missing package details input', () => {
    // TODO: TEST: should throw error on missing package details input
  });

  it('should throw error when twitter get credentials throws an error', async () => {
    expect.assertions(1);

    twitterMocks.reset({ get: (...args: any) => Promise.reject('Twitter Get Error') });

    try {
      await announceIt.announceRelease(packageDetails);
    } catch (e) {
      // TODO: FEATURE: create a basic KbError class
      expect(e).toMatch('Twitter Get Error');
    }
  });

  it('should throw error when twitter post tweet throws an error', async () => {
    // TODO: TEST: should throw error when twitter post tweet throws an error
  });

  it('should throw error when using missing variables in template', async () => {
    // TODO: TEST: should throw error when using missing variables in template
  });
});
