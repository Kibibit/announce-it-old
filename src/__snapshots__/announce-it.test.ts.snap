// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KbAnnounceIt should create instance when given correct parameters 1`] = `
KbAnnounceIt {
  "client": Object {
    "get": [Function],
    "post": [Function],
  },
}
`;

exports[`KbAnnounceIt should throw error when missing options 1`] = `"ERROR: missing required options"`;

exports[`KbAnnounceIt should throw error when missing options 2`] = `"ERROR: missing required options"`;

exports[`KbAnnounceIt should throw error when missing options 3`] = `"ERROR: missing required options"`;

exports[`KbAnnounceIt should throw error when missing options 4`] = `"ERROR: missing required options"`;

exports[`KbAnnounceIt should throw error when missing options 5`] = `"ERROR: missing required options"`;

exports[`kbAnnounceIt.announceRelease should post to twitter when stable release 1`] = `"test-template"`;

exports[`kbAnnounceIt.announceRelease should post to twitter when unstable release and mentioned in packageDetails 1`] = `"test-template"`;

exports[`kbAnnounceIt.announceRelease should throw an error when unstable release and not mentioned in packageDetails 1`] = `"Not a stable release"`;

exports[`kbAnnounceIt.announceRelease should throw error when twitter get credentials throws an error 1`] = `"Twitter Get Error"`;

exports[`kbAnnounceIt.announceRelease should throw error when twitter post tweet throws an error 1`] = `"Twitter Post Error"`;

exports[`kbAnnounceIt.generateTweet should throw error on missing package details input 1`] = `"The key name is missing from given object"`;

exports[`kbAnnounceIt.generateTweet should throw error on non object input 1`] = `"Expecting Object"`;

exports[`kbAnnounceIt.generateTweet should throw error when using missing variables in template 1`] = `"Using missing variables in tweet template"`;
